src/frontend/src/components/tables/InvenTreeTable.tsx

```diff
@@ -0,0 +1,347 @@
+import { t } from '@lingui/macro';
+import { DataTable, DataTableSortStatus } from "mantine-datatable";
+
+import { useEffect, useState } from 'react';
+import { useQuery } from '@tanstack/react-query';
+import { api } from '../../App';
+
+import { ActionIcon, Chip, CloseButton, Indicator, Space, Stack, Table, Text, Tooltip, filterProps } from '@mantine/core';
+import { Group } from '@mantine/core';
+
+import { TableColumnSelect } from './ColumnSelect';
+import { TableSearchInput } from './Search';
+import { DownloadAction } from './DownloadAction';
+
+import { IconFilter, IconRefresh } from '@tabler/icons-react';
+
+import { IconBarcode, IconPrinter } from '@tabler/icons-react';
+
+import { ButtonMenu } from '../items/ButtonMenu';
+import { notYetImplemented } from '../../functions/notifications';
+import { FilterGroup } from './FilterGroup';
+
+import { TableColumn } from './Column';
+import { TableFilter } from './Filter';
+
+/**
+ * Table Component which extends DataTable with custom InvenTree functionality
+ */
+
+export function InvenTreeTable({
+    url,
+    params,
+    columns,
+    enableDownload=false,
+    enableFilters=true,
+    enablePagination=true,
+    enableRefresh=true,
+    enableSearch=true,
+    enableSelection=false,
+    pageSize=25,
+    tableKey='',
+    defaultSortColumn='',
+    noRecordsText=t`No records found`,
+    printingActions=[],
+    barcodeActions=[],
+    customActionGroups=[],
+    customFilters=[],
+} : {
+    url: string;
+    params: any;
+    columns: TableColumn[];
+    tableKey: string;
+    defaultSortColumn?: string;
+    noRecordsText?: string;
+    enableDownload?: boolean;
+    enableFilters?: boolean;
+    enableSelection?: boolean;
+    enableSearch?: boolean;
+    enablePagination?: boolean;
+    enableRefresh?: boolean;
+    pageSize?: number;
+    printingActions?: any[];
+    barcodeActions?: any[];
+    customActionGroups?: any[];
+    customFilters?: TableFilter[];
+}) {
+
+    // Check if any columns are switchable (can be hidden)
+    const hasSwitchableColumns = columns.some((col: any) => col.switchable);
```

Comment:
For performance reasons you could wrap this in a useMemo here and all the functions in a useCallback react hooks, arrow functions also look much clear and you can save the {} here.

Reply:
```ts
import {useMemo} from "react";
const hasSwitchableColumns = useMemo(() => columns.some((col: any) => col.switchable), [columns]);
```

And for an useCallback example:
```ts
const updateSearchTerm = useCallback((term: string) => {
        term = term.trim();
        // Ignore identical search terms
        if (term == searchTerm) return;

        setSearchTerm(term);
        latestSearchTerm = term;
        refetch();
}, [searchTerm, refresh]
```
You can read more on the [react docs](https://react.dev/reference/react/useCallback) page.